runtime/parser2/lexer/lexer.go

```diff
@@ -95,15 +96,35 @@ func (l *lexer) Revert(cursor int) {
 	l.cursor = cursor
 }
 
+func (l *lexer) clear() {
+	l.startOffset = 0
+	l.endOffset = 0
+	l.prevEndOffset = 0
+	l.current = EOF
+	l.prev = EOF
+	l.canBackup = false
+	l.startPos = position{line: 1}
+	l.cursor = 0
+	l.tokens = l.tokens[:0]
+	l.tokenCount = 0
+}
+
+func (l *lexer) Reclaim() {
+	pool.Put(l)
+}
+
+var pool = sync.Pool{
+	New: func() interface{} {
+		return &lexer{
+			tokens: make([]Token, 0, 2048),
+		}
+	},
+}
+
 func Lex(input string) TokenStream {
-	l := &lexer{
-		input:         input,
-		startPos:      position{line: 1},
-		endOffset:     0,
-		prevEndOffset: 0,
-		current:       EOF,
-		prev:          EOF,
-	}
+	l := pool.Get().(*lexer)
+	l.clear()
```

Comment:
Wouldn't it be better for the `Reclaim()` to do the clearing (i.e: `l.clear()`) before putting it back in the pool? Then the memory (e.g: tokens array) is freed up while the lexer is in the pool.

Reply:
But wouldn't that defeat the optimization, re-using the tokens array? Allocating the lexer itself isn't the expensive part, it is the tokens array in it.

Reply:
ah right, the array isn't cleared. I misread the 'slicing' as 'array re-initialization'.

Reply:
But yeah, I think the clearing / initialization is better to do after getting the item from the pool, for two reasons:
- It mimics initialization like the "construction" would do
- The returned item may not be a recycled one, but a newly constructed one (the "construction function" passed to the the pool), so we would have to call clear/init in two places: when the lexer is constructed, and when it is to be reclaimed